### ISSUE TITLE
Inconsistent error stack in standalone mode

### ISSUE THREAD

**Issue by jtlapp** [webdriverio/webdriverio#1609]
OS:  Mac OS X 10.11.4 x86_64
Node: v6.3.1
Selenium Version:  v2.53.1, with Core v2.53.1. Built from revision a36b8b1
webdriver.io: 4.2.16

When a callback provided to `webdriver.io` throws an exception or returns a rejected promise, the client properly bubbles up the error, except when chaining `.end()`.

I tested errors generated as follows:
- `throw new Error("an exception happens here")`
- `return Promise.reject(new Error("a promise gets rejected"))`
- `return Promise.reject("a promise gets rejected")`

Here is a demonstration of throwing an exception that works properly:

``` js
var promise = webdriverio
    .remote(options).init()
    .url('http://www.google.com')
    .getTitle().then(function(title) {
        throw new Error("an exception happens here");
        console.log('Title was: ' + title);
    });
    promise.end();
showError(promise);

function showError(promise) {
    if (typeof promise.then === 'function') {
        promise.catch(function (err) {
            if (err instanceof Error)
                console.log(err);
            else
                console.log("Rejected promise: "+ err);
        });
    }
    else
        console.log("not a promise");
}

// Error: an exception happens here
//     at Object.<anonymous> (/repos/arepo/tests/test_errs.js:26:15)
//     at Object.exec (/repos/arepo/node_modules/webdriverio/build/lib/helpers/safeExecute.js:28:24)
//     at Object.resolve (/repos/arepo/node_modules/webdriverio/build/lib/webdriverio.js:189:29)
```

If we instead return the result of chaining `.end()`, we get this:

``` js
var promise = webdriverio
    .remote(options).init()
    .url('http://www.google.com')
    .getTitle().then(function(title) {
        throw new Error("an exception happens here");
        console.log('Title was: ' + title);
    })
    .end();
showError(promise);

// Error: an exception happens here
//     at title() - getTitle.js:29:28
//     at end() - webdriverio.js:16:6
```

For completeness, here are comparisons of chaining vs not chaining returning rejected promises:

``` js
        return Promise.reject(new Error("a promise gets rejected"));

/* NOT CHAINED */
// Error: a promise gets rejected
//     at Object.<anonymous> (/repos/arepo/tests/test_errs.js:56:31)
//     at Object.exec (/repos/arepo/node_modules/webdriverio/build/lib/helpers/safeExecute.js:28:24)
//     at Object.resolve (/repos/arepo/node_modules/webdriverio/build/lib/webdriverio.js:189:29)

/* CHAINED */
// Error: a promise gets rejected
//     at title() - getTitle.js:29:28
//     at end() - test_errs.js:45:6
```

``` js
        return Promise.reject("a promise gets rejected");

/* NOT CHAINED */
// Rejected promise: a promise gets rejected

/* CHAINED */
// (nothing is output in this scenario)
```

I suspect a decision must be made about whether `.end()` should return a promise that propagates a callback error. If the answer is yes, it should be made consistent. If the answer is no, `.end()` should return no error information instead of a confused mix of information. The inconsistencies had me pulling my hair out looking for the proper resolution.

My preference is to be able to include include the following in my tests and have it properly propagate errors to the test harness:

``` js
var t = require('tap');

t.test("load the home page", function (t) {
    return webdriverio
        .remote(options).init()
        .url('http://www.google.com')
        .getTitle().then(function(title) {
            console.log('Title was: ' + title);
        })
        .end();
});
```

Note that [`tap`](https://github.com/tapjs/node-tap) will handle properly returned promises in these error scenarios, reporting the errors.

**Comment by jtlapp** [comment]
Things got stranger. My solution doesn't quite work. If I wrap the error within another promise, the error again reverts to gobbledygook. The following code:

```js
var promise = new Promise(function (resolve, reject) {
    var p = webdriverio
        .remote(options).init()
        .url('http://www.google.com')
        .getTitle().then(function(title) {
            throw new Error("an exception happens here");
            console.log('Title was: ' + title);
        });
    p.end();
    resolve(p);
});
showError(promise);
```

Produces this incorrect output:

```
Error: an exception happens here
    at title() - getTitle.js:29:28
    at end() - test_errs.js:113:
```

Recall that if I don't wrap it within an additional promise, I get the correct output:

```
Error: an exception happens here
    at Object.<anonymous> (/repos/arepo/tests/test_errs.js:26:15)
    at Object.exec (/repos/arepo/node_modules/webdriverio/build/lib/helpers/safeExecute.js:28:24)
    at Object.resolve (/repos/arepo/node_modules/webdriverio/build/lib/webdriverio.js:189:29)
```

Any idea what's going on here?

**Comment by jtlapp** [comment]
`error.stack` is getting overwritten. If I save the correct stack in `error.correctStack` and print it out later, it remains correct, despite an incorrect `error.stack`:

```js
var promise = new Promise(function (resolve, reject) {
    var p = webdriverio
        .remote(options).init()
        .url('http://www.google.com')
        .getTitle().then(function(title) {
            var e = new Error("an exception happens here");
            e.correctStack = e.stack;
            throw e;
            console.log('Title was: ' + title);
        });
    p.end();
    resolve(p);
});
```

**Comment by jtlapp** [comment]
I have a new workaround. Consider the following, which wraps the call to the webdriver client in a `browse()` function:

```js
var promise = new Promise(function (resolve, reject) {
    // do the first asynchronous job
    resolve();
})
.then(function () {
    return browse(webdriverio.remote(options).init()
        .url('http://www.google.com')
        .getTitle().then(function(title) {
            throw new Error("an exception happens here");
            console.log('Title was: ' + title);
        })
    )
});
showError(promise);
```

If we define `browse()` as follows, we preserve the stack trace:

```js
function browse(client) {
    client.end();
    return client.catch(function (err) {
        // preserve stack, which otherwise gets overwritten
        var newError = new Error(err.message);
        newError.stack = err.stack;
        throw newError;
    });
}

// Error: an exception happens here
//     at Object.<anonymous> (/repos/arepo/tests/test_errs.js:126:19)
//     at Object.exec (/repos/arepo/node_modules/webdriverio/build/lib/helpers/safeExecute.js:28:24)
//     at Object.resolve (/repos/arepo/node_modules/webdriverio/build/lib/webdriverio.js:189:29)
```

If we simplify `browse()` as follows, we get the mucked up stack trace, but the browser window still closes on error:

```js
function browse(client) {
    client.end();
    return client;
}

// Error: an exception happens here
//     at title() - getTitle.js:29:28
//     at end() - test_errs.js:145:16
```

If `browse()` only calls `client.end()`, we still get the mucked up stack trace, and the browswer window remains open too:

```js
function browse(client) {
    return client.end();
}

// Error: an exception happens here
//     at title() - getTitle.js:29:28
//     at end() - test_errs.js:145:16
```

It was a struggle to find a version of `browse()` that preserved the stack trace in all circumstances, and I don't have high confidence that this solution will always work. I'd still like a more robust solution.

**Comment by jtlapp** [comment]
I believe I have resolved this problem. It's a consequence of misunderstanding the documentation and mimicking poor examples found in the documentation. I now think I can write the missing webdriver.io doc page that crucial to a basic understanding of this API. Stay tuned.

**Comment by jtlapp** [comment]
If there is a bug, I believe it is that `.end()` and `.endAll()` should only be available on the immediate output of `.remote()`. Applying it anywhere down the chain obfuscates errors.

**Comment by jtlapp** [comment]
And if you want to improve the clarity of the API, rename `init()` to `openWindow()` and `end()` to `closeWindow()`. Unless I'm still confused.

**Comment by jtlapp** [comment]
Drats. Despite my supposed greater understanding, the stack trace still sometimes gets munged.

**Comment by jtlapp** [comment]
This seems to be the most reliable fix:

```js
var client = webdriverio.remote(options);
var promise = client.init()
    .url('http://www.google.com')
    .getTitle().then(function(title) {
        fix(function () {
            throw new Error("an exception happens here");
            console.log('Title was: ' + title);
        });
    })
    .end()
    .catch(function (err) {
        client.end();
        rethrow(err);
    });
showError(promise);

function fix(func) {
    try {
        return func();
    }
    catch (err) {
        err.correctStack = err.stack;
        throw err;
    }
}

function rethrow(err) {
    err.stack = err.correctStack;
    delete err.correctStack;
    throw err;
}
```

**Comment by jtlapp** [comment]
All this effort on webdriver.io, and I now think I have to abandon it. I insist on using my own test runner, and I have no idea how to debug webdriver.io without wdio. stdout appears to go to the bitbucket, so I can't write progress logs. The debugger statement is ignored, presumably because it's running commands in a separate process. Any reason not to pipe child stdout to the parent?

**Comment by christian-bromann** [comment]
> I insist on using my own test runner

What requirements do you have? Why is the wdio test runner not an option for you?

> stdout appears to go to the bitbucket

What do you mean by that?

> The debugger statement is ignored, presumably because it's running commands in a separate process.

No, it is due to the async nature of chaining these calls. The debug statement works great if you have a collection of synchronous calls but the way how WebdriverIO allows to chain async calls in order to provide some kind of sync nature doesn't work with that. However the wdio testrunner allows to use the [debugging statement](http://webdriver.io/guide/testrunner/debugging.html) as it calls were transformed to work synchronous.

> Any reason not to pipe child stdout to the parent?

I guess you are referring to why does WebdriverIO in standalone mode not throw an error. Simply because it handles commands as promises. A command that fails returns a rejected promise and all commands that are chained after it will be rejected too.

**Comment by jtlapp** [comment]
Many factors go into choosing a test runner. One of the most significant is the underlying test harness, such as `mocha`, `tape`, or [`tap`](http://www.node-tap.org/api/). I believe `tap` to be the most robust test harness available, particularly after its rewrite last year. It just needs a tool for debugging tests. I'm optimizing [`subtap`](https://github.com/jtlapp/subtap) for that. This can't be done generically for all possible test harnesses.

So I've already chosen the test runner I intend to use for debugging `tap` tests. I'm now looking to make it ideal for debugging `tap` tests with Selenium. I need a library for driving Selenium that is test-runner agnostic. But thank you for the education along the way.

**Comment by jtlapp** [comment]
> stdout appears to go to the bitbucket

Correction: stdout and stderr are lost when using the `tap` test runner. I've never seen that before in `tap`, but it's possible that this could be a `tap` bug.

**Comment by christian-bromann** [comment]
@jtlapp being able to choose the underlying test harness was always important when I implemented wdio. That is why you can choose between Mocha, Jasmine and Cucumber as well as any arbitrary assertion library. We haven't build an adapter for tape or tap yet but consider doing it. However you need to understand that running unit test with tap and tape is a lot different than running integration tests with Selenium:

1. unit test don't need any preparation, you just import your function and test it, selenium tests need a running selenium session as well as a running test environment
2. unit tests run sequentially, Selenium tests don't - if you want to scale your Selenium test suite you need to run your test in parallel because Selenium tests take way longer than unit test
3. reporting: unit test framework reporter are not made for running selenium tests - if you run into issues you look for different things than the only the exception, wdio reporters help you to find the root of a failing test quickly by providing screenshots and other logs

I understand that you prefer running tests with tape or tap but keep in mind that the framework itself is not necessary appropriate for running Selenium tests. I would recommend to go with wdio and mocha. At least just try it out. The test runner comes with a lot of important integrations you will need if you scale your test suite.

**Comment by jtlapp** [comment]
Thanks for the thoughtful response. (1) is a non-issue. `tap` and `subtap` deal with this easily. (2) and (3) are great points. I don't see why we can't expand `tap` to be able to run at least separate test files asynchronously, and that would suffice. You'd only ever want `subtap` running tests sequentially because you use it to debug. (3) is a more serious issue and the one that may yet bring me back, but I would first be looking at extending `subtap` to support Selenium.

Now that I'm returning to selenium-webdriver, I find it quite ugly and bulky. You've spoiled me!

**Comment by christian-bromann** [comment]
I really don't understand why you are oppose to use wdio in combination with mocha? Everything you get from it like reporters (e.g. [sumologic reporter](https://github.com/webdriverio/wdio-sumologic-reporter)), services (e.g. like the [sauce service](http://webdriver.io/guide/services/sauce.html)) and synchronous commands outweigh all robustness that tap or subtap has more compared to mocha. Just want you to keep that in mind when switching back 😉

**Comment by jtlapp** [comment]
A good question, and one I've wondered about in retrospect. I'm an old-school software developer, been coding since 1982, following all the fads. However, I'm relatively new Javascript and app dev with Javascript. When I went shopping for a test framework and saw mocha, a loud alarm sounded:

Mocha leverages Javascript to create an English-like syntax. This makes for quite readable code. But I've seen this done many times before. I've seen people make Prolog English-like, use overloading to make C++ English like, and more recently Ruby/Rails strives to define a declarative language out of terms. I've also done all this myself. The effect has always always been the same: it is harder to write and debug reliable code, because the declarations hide way too much behavior. I have had to learn this lesson over and over, and now the lesson is learned.

More simply, Mocha does not meet my criteria for good design. (Never mind the business of using globals that everyone complains about. Design needs to be evaluated for its effect, not on abstract principles.)

That said, Mocha has a lot of weight and utility behind it by now, so there's great incentive to live with the pain of bad design in order to keep from having to reinvent the wheel.

So I went searching for an alternative framework. I cannot recall finding a single page that compared Mocha variably to Tape. The preference was always greatly for Tape. Here are some of the links I encountered:

- https://medium.com/javascript-scene/why-i-use-tape-instead-of-mocha-so-should-you-6aa105d8eaf4#.rn2zru5f1 
- http://karloespiritu.com/better-unit-testing-with-tape/ 
- https://remysharp.com/2015/12/14/my-node-test-strategy 
- http://www.macwright.org/2014/03/11/tape-is-cool.html
- http://substack.net/how_I_write_tests_for_node_and_the_browser 
- https://ponyfoo.com/articles/testing-javascript-modules-with-tape 

I went off using tape and discovered that it was missing some features I preferred having when debugging. I tried to monkey patch it. That didn't work. I tried to fork and extend it ([`tapeo`](https://github.com/jtlapp/tape)), but that didn't work either because the Tape architecture is too fragile.

Then I discovered [`tap`](https://github.com/tapjs/node-tap), written by the inventor and CEO of NPM and the `npm` command, completely rewritten last year for greater flexibility and robustness. The more I work with it, the more impressed I am. It is a great tool, and it is the tool that tests should be using, although there's no option for running tests concurrently at the moment. I found that I could easily extend it with simple monkey-patching ([`subtap`](https://github.com/jtlapp/subtap)) and so have been trying to help right the direction of Javascript testing. Not that I don't have far more important things to do.

Well, that was good to get in writing. But that's my explanation.

**Comment by christian-bromann** [comment]
I see. Thanks for the explanation.

Have you checked out our [page object example](https://github.com/webdriverio/webdriverio/tree/master/examples/pageobject) (or [this](http://webdriver.io/guide/testrunner/pageobjects.html) page)? This is how many WebdriverIO user write their e2e tests and do like it. I agree on the points you mentioned above on Mocha but starting monkeypatching a different framework also don't seem to be a good solution. At the end of the day it matters how quick and stable you can write your e2e tests if with this solution (wdio + page objects + mocha) you can do so at high scale. Going back to something that runs commands on promises will introduce race conditions pretty quickly also building the whole setup by yourself and maintaining it also eats up a lot of time. We will look into an integration with tape asap but it wouldn't change much in regards of the test procedure.

**Comment by jtlapp** [comment]
I'll have a look. Thanks. "Monkey patching" might not be the right term for what I did. Still new to JS. I'm overloading public APIs. In OO languages you would do that by extending a class. In JS I can do that by reaching into the existing class. So maybe "extension" is a better term. So far all the occasions were I felt the need to reach into `tap` internals have resolved without that need. I recently added one internal dependency that I may be able to resolve the way others have.

**Comment by jtlapp** [comment]
I've made a discovery that makes it possible for me to do simple debugging using `tap`: I can output to stdout from within Promise::catch() after a Selenium-sourced error.

So I created a `logLine` command that writes lines to a string. I dump the contents of that string to stdout from my `rethrow()` function, clearing the string. I also dump this string on exiting the process. Now at least I can add logs that allow for locating the problem command.

Should quick-and-dirty debugging not suffice for a particularly problematic run, we can see what `wdio` tells us. Now we can remain in `tap` for most of our testing.

It is *very* hard to return to selenium-webdriver. webdriverio tests are easy to understand at a glance, and easy to write too. But I would still like better support for `tap` and for 3rd-party debugging, as well as fixes to these strange error problems.

Thanks for your patience with me!

**Comment by christian-bromann** [comment]
From what I see is that the central problem is the different error stack you get when ending the chain with `end()` or not. I kind of scares me that you want to end your selenium session after each test bit. However you are right that this is not consistent but also doesn't have high priority for me (PRs welcome though). I still don't get why you've choosen selenium-webdriver over WebdriverIOs offering after everything I said 😆 .

**Comment by jtlapp** [comment]
I didn't follow that. I found a way to use WebdriverIO with `tap`, so I'm sticking with WebdriverIO.

I don't know what you mean by ending the "selenium session after each test bit". I'm running multiple distinct tests within a single file. Each test is a single promise chain.

I seem to be dealing with two errors:

1. The stack trace of an exception thrown within a `.then` gets munged if any command at all follows the `.then`. If all that follows is a `.catch`, the stack trace is fine.
2. If a Selenium-sourced error (an error reported by the server to WebdriverIO) occurs during a test run, everything written to stdout vanishes. I can't even collect it into a string for writing in a process exit handler. However, I can write to stdout within the Promise::catch() that catches the exception. Everything written to stdout prior is gone, so I write my collected output anew.

**Comment by jtlapp** [comment]
Woo! I've written a custom command that pauses testing from `subtap` until the user hits ENTER. Works fine because `subtap` necessary runs tests sequentially for debugging purposes.

**Comment by jyoungqu** [comment]
I've ran into this problem as well.  I am, but will no longer, using `browsermob-proxy` which uses `webdriverio` to pull site pages and test on network traffic captured in HAR.  After about 20-25 pages (out of 22K) I'll get the `Error:  couldn't connect to selenium server at end() - index.js:38`  I'll have to move on to use something I can trap the errors on and pick up with my page crawling.

**Comment by christian-bromann** [comment]
@jyoungqu your problem is not really related to this issue

**Comment by jyoungqu** [comment]
@christian-bromann the selenium server never went away.  But probably somewhat correct that possibly this isn't a related issue.  The only reason I posted on this is that better error trapping would be nice to have to recover rather than just exiting out of the script, I've tried a few options to catch the error but don't have time to go as deep as this tread has.

**Comment by christian-bromann** [comment]
@jyoungqu each command represents a promise, so you can always react on errors. What should WebdriverIO instead do if it comes across an issue like that? It can't recover just like that. Especially when you chain commands like in standalone mode.

**Comment by jtlapp** [comment]
The error I described for this issue originates from code that you put within a `.then`, and its manifestation is only an incorrect stack trace in the exception. It's interim solution is [here](https://github.com/webdriverio/webdriverio/issues/1609#issuecomment-249784165).

**Comment by realzoberg** [comment]
I found that Q already got option called Q.longStackSupport. Can anybody advice how enable it in webdriver.io?

**Comment by CSomerville** [comment]
This has come up for me repeatedly, and is a feature I would love to have. Would you accept this as a flag passed to remote, eg `{ verboseStackTraces: true }`? If so, `Q.longStackSupport` (which imposes performance penalties but is useful for debugging) could be enabled and the call to `Error.captureStackTrace()` bypassed. Let me know if a solution along these lines would be amenable to you. Many thanks for your time and effort.

**Comment by CSomerville** [comment]
Nevermind about bypassing `Error.captureStackTrace()`, looks like it is as simple as enabling `q.longStackSupport`.

**Comment by hackhat** [comment]
I'm using await and I got only the message `Error` in the stack:

Code:

    try {
      await this.client.click(`...`);
    } catch (e) {
      console.log(`>>>1`, e); // >>>1 { Error message: 'unknown error: Element is...
      console.log(`>>>2`, e.stack); // >>>2 Error
      console.log(`>>>3`, e.message); // >>>3 unknown error: Element is not clickable at point (400, 162). Other element would receive the click: ...
      throw e;
    }


` >>>2` log is incomplete, is missing the whole stack;

Don't know if this helps.

**Comment by franciskim** [comment]
@jtlapp I think you could try a:

```
.then()
.catch()
.end()
.catch()
```

pattern

**Comment by christian-bromann** [comment]
Closing as this will be obsolete in v5.
